[ BB2-805 ] BFD Insights - Implement global-state firehose delivery

dtisza1 opened the pull request:
<!--
You've got a Pull Request you want to submit? Awesome!
This PR template is here to help ensure you're setup for success:
  please fill it out to help ensure that your PR is complete and ready for approval.
-->

**JIRA Ticket:**
[BB2-805](https://jira.cms.gov/browse/BB2-805)
[BB2-130](https://jira.cms.gov/browse/BB2-130)
[BB2-777](https://jira.cms.gov/browse/BB2-777)


**User Story or Bug Summary:**
<!-- Please copy-paste the brief user story or bug description that this PR is intended to address. -->
For one BB2 log event, make an inital transport for a BB2 log events out to a Kinesis Firehose delivery stream in preparation for integration in to BFD-insights.

Summary:

  * This will send logging events from the nightly apps/logging/management/commands/log_global_state_metrics.py management command to the BFD-Insights Kinesis Firehose streams named "bfd-insights-bb2-global-state" and "bfd-insights-bb2-global-state-apps".

  * Kept the Kinesis Firehose logging separate from the Python logger to simplify. For logging integration we will be looking in to utilizing either CloudWatch or a Firehose agent. This is for an asynchronous type solution vs. one that can block the BB2 web services. This method is good for the nightly global state type events and keeps it simple for now.

  * Reorganized the docker-compose setup related to sourcing ENV variables and for the ability to test Djano management commands. Also updated the local development readme.md based on this.

  * Added code to the server startup (base.py) that gets the image & instance ids from the instance's metadata endpoint. This can later be used in our python logging. It is currently just used for the Firehose events.

  * Documentation on the setup is available in confluence under our "BFD-Insights" section. For example, this is working with a BB2 cross account role.


### What Does This PR Do?

<!--
Add detailed description & discussion of changes here.
The contents of this section should be used as your commit message (unless you merge the PR via a merge commit, of course).

Please follow standard Git commit message guidelines:
* First line should be a capitalized, short (50 chars or less) summary.
* The rest of the message should be in standard Markdown format, wrapped to 72 characters.
* Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behavior.
* List all relevant Jira issue keys, one per line at the end of the message, per: <https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html>.

Reference: <https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project>.
-->
**The following is a list of changes from the commits (including additional comments):**


NOTE: There are lots of changes in the commit history while performing discovery. Removed them from this section after similifying the code.


* Kinesis Firehose setup
  - Review the changes in the following files:
  ``` 
  hhs_oauth_server/settings/base.py
  apps/logging/firehoses.py
  apps/logging/loggers.py

  ``` 


* EC2 instance metadata:
  - Review the changes in the following files:
  ```
  hhs_oauth_server/settings/base.py
  hhs_oauth_server/utils.py
  ```

* Docker Compose ENV settings reorganization:
  - Review the changes in the following files:
  ```
  hhs_oauth_server/settings/base.py
  docker-compose.yml
  docker-compose/bluebutton_server_start.sh
  docker-compose/db-env-vars.env
  docker-compose/msls-env-vars.env
  docker-compose/readme.md
  docker-compose/slsx-env-vars.env
  docker-compose/unittests-env-vars.env
  docker-compose/web-env-vars.env
  ```



### What Should Reviewers Watch For?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply for some reason.

Common items include:
* Is this likely to address the goals expressed in the user story?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Is user input properly sanitized & handled?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
-->
If you're reviewing this PR, please check these things, in particular:

* Are there any unhandled and/or untested edge cases you can think of?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
* Naming in general? Do the namings of variables, methods, classes, and other items make sense? Are they too verbose, or not verbose enough? Any redundancies? Are the comments descriptive enough?
* Are audit logging items properly handled? Is sensitive information either not included or hashed?

TESTING SCRIPT:

The following testing was performed:

* Followed the updated instruction for starting the web container with AWS, BFD and SLSx settings/credentials in docker-compose/readme.md

* To enable the Firehose, add the following to your `.env` file in the project root. See the configuration doc in confluence for the role info.
  ```
  DJANGO_LOG_FIREHOSE_ENABLE=true
  DJANGO_LOG_FIREHOSE_CROSSOVER_ROLE_ARN=<arn for crossover role>
  ```

* Source the BB2 AWS environment variables in to your terminal/shell session per the method for your OS.

* Test that events for the `log_global_state_metrics()` function are routed through the two Kinesis Firehose delivery streams and that the results end up in the target S3 bucket paths.
  - This is run using the following Django management command:
  ``` 
  docker-compose exec web python manage.py log_global_state_metrics
  ``` 
  - Tested by generating a large number of applications locally.
  - Edited source to produce the different types of BFDInsightsFirehoseException exceptions in apps/logging/firehoses.py.

* NOTE: If you are getting AWS token related errors, you will need to restart everything as follows to source the new token in to the docker container. Also follow the AWS credentials sourcing instruction in the local development instructions.

  ```
  docker-compose down
  docker-compose up -d
  docker-compose exec web docker-compose/migrate.sh
  ```


### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies?  **No**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **No**.


### What Needs to Be Merged and Deployed Before this PR?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply.

Common items include:
* Database migrations (which should always be deployed by themselves, to reduce risk).
* New features in external dependencies (e.g. BFD).
-->

The following bluebutton-web-deployment repo PR needs to be merged before this will work for deployments: https://github.com/CMSgov/bluebutton-web-deployment/pull/1444 


### Any Migrations?

<!--
Make sure to work with whoever is doing the deploy so they are aware of any migrations that may need to be run
-->

* [ ] Yes, there are migrations
  * [ ] The migrations should be run PRIOR to the code being deployed
  * [ ] The migrations should be run AFTER the code is deployed
  * [ ] There is a more complicated migration plan (downtime, etc) <!-- Make sure to include the details of the plan below -->
* [x] No migrations


### Submitter Checklist

<!--
Helpful hint: if needed, Git allows you to edit your PR's commits and history, prior to merge.
See these resources for more information:

* <https://dev.to/maxwell_dev/the-git-rebase-introduction-i-wish-id-had>
* <https://raphaelfabeni.com/git-editing-commits-part-1/>
-->

I have gone through and verified that...:

* [ X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [ X] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [X ] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X ] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X ] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X ] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [X ] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

nbragdon commented on apps/health/checks.py:
```diff
@@ -38,6 +39,12 @@ def slsx(v2=False):
     return slsx_client.service_health_check(None)
 
 
+def firehose_check(v2=False):
```
Nice addition to the health checks 👍

nbragdon commented on apps/logging/firehoses.py:
```diff
@@ -0,0 +1,82 @@
+            self.__firehose = boto3.client('firehose')
+        except Exception as err:
+            mesg = "Firehose client initialization failed. err: " + str(err)
+            raise BFDInsightsFirehoseException(mesg)
```
Would this result in a 500 response for users?

dtisza1 replied:
Good call out!   I'll perform some more testing with this to verify.

nbragdon commented on docker-compose.msls.yml:
```diff
@@ -0,0 +1,77 @@
+    build: . 
+    command: ./docker-compose/bluebutton_server_start.sh '${DB_MIGRATIONS}' '${SUPER_USER_NAME}' '${SUPER_USER_EMAIL}' '${SUPER_USER_PASSWORD}' '${BB20_ENABLE_REMOTE_DEBUG}' '${BB20_REMOTE_DEBUG_WAIT_ATTACH}'
+    environment:
+            - DJANGO_SETTINGS_MODULE=hhs_oauth_server.settings.dev
```
For env vars, you can actually add them to a file and just use the file name here if that is easier for organization.

nbragdon replied:
Ex:
```
web:
  env_file:
    - web-variables.env
```
You can read more on it here: https://docs.docker.com/compose/environment-variables/

dtisza1 replied:
We could.  I feel that some of the ENVs are better to have sourced vs. stored locally.  Deployment (Jenkins jobs) would likely need updated to work with this also. This might be a good improvements story for the backlog?

nbragdon replied:
I don't think it would have an impact on deployment. Could have @oragame or @njdister weigh in though. Basically the deployment jobs checkout the repo, and then run the docker-compose from there. You'd basically create the `.env` in the repo and check it in (since the values are already checked into the repo they are safe to be in there). It's just basically more organized to keep them in a separate file to really keep the docker-compose specific to the setup of the containers.

nbragdon replied:
Basically can we do: `AWS_DEFAULT_REGION=${AWS_DEFAULT_REGION-us-east-1}` inside the `.env` file and it work the same?

dtisza1 replied:
@nbragdon I'm liking this idea more after discussing with you in more detail. Will add this in :smile_cat:

oragame replied:
My .02 as far as .env's - I think in this particular case there is some repetition of environment vars on web and unittests in the dc file, so there could be some organization around that with .env files. I am thinking maybe a .common.env for the values used on both services, and then a specific one for web like .web.env - sounds like you have a plan, good call Nick!

dtisza1 replied:
@nbragdon @oragame I just reworked the ENV configs. This tested out OK and looks cleaner.  Split the env files up by container and SLSx/MSLS. If having any issues, there is sometimes the need to docker-compose down/up to get changes to register.

@JFU-GIT I removed the shell arguments passed to the bluebutton_server_start.sh script. Hoping this doesn't break any functionality. In testing, the ENVs worked if sourced from a developer's .env or launching shell OK.

nbragdon commented on hhs_oauth_server/settings/base.py:
```diff
@@ -379,6 +408,10 @@
             'level': 'INFO',
         },
+        'audit.global_state_metrics': {
+            'handlers': ['console', 'kinesis'],
```
Nice 👍 , this looks like the ideal way to handle this

nbragdon replied:
This doesn't need to change but just a thought: how are the logs getting to splunk? Is it basically collecting after it goes to the console?

dtisza1 replied:
@nbragdon That's right. Splunk has a log ingestion agent that runs on the instances. It gets the logging directly from the log files.

nbragdon replied:
Just going to add in here for reference: The handlers don't seem to be async and if there is a long timeout or some other issue with the kinesis stream it'll hold up processing user requests. So we are looking at other options, either cloudwatch logs or an agent that pulls from the log files instead.

dtisza1 replied:
+1 for this. The log handlers are blocking (not async), so just going to use this method related to the "audit.global_state_metrics" type logging for now. Those run on a nightly schedule via Jenkins & Django management command---so will not impact the web services, if they have issues.  Going to refactor that logging with this in mind next...

nbragdon commented on hhs_oauth_server/utils.py:
```diff
@@ -1,9 +1,15 @@
+import logging
```
Lets use the new logger here :) 
```suggestion
import apps.logging.request_logger as logging
```

nbragdon commented:
@dtisza1 Overall this is looking good. Nice work here

dtisza1 commented:
@nbragdon @oragame @JFU-GIT This is ready for review related to just the BB2-777 Jira story (and the docker-compose changes). This was just related to getting events in to the BFD-Insights Kinesis Firehose from local development.

This PR will continue related to BB2-805 for any changes needed for this to run via the nightly Jenkins "RUN - Global State Metrics (env)" jobs. So will not merge this in until that work is added. 

See the updated PR summary and testing info, since this has change a lot. I simplified the Firehose class to concern just what is needed for the nightly global state metrics events, without tying in to the python logging. We will work on an async solution for other types of logging.  I also removed the health-check, since it is not needed for this purpose.

TY!

JFU-GIT reviewed (approved):
Looks good, tested and worked as expected

oragame commented:
Updates are looking really good since my last review on this, Dave. Nice work!

dtisza1 commented:
@oragame @nbragdon @JFU-GIT This is ready for review related to updates per the BB2-805 story AC.

I was able to successfully test this via the "RUN - Django Command" project/job in Jenkins CBC using a TEST canary.  See the related deployment repo PR for more details at:  https://github.com/CMSgov/bluebutton-web-deployment/pull/1444  .  You can view the console log for build #420, if you want to just view the results.

If you run in to permissions issues when testing locally, your AWS user account might need to be added to the cross-over role setup in BB2 AWS IAM.  See the confluence docs for more details about this or let me know and I can add you.

Thank you!

oragame commented on hhs_oauth_server/settings/base.py:
```diff
@@ -325,6 +326,25 @@
+TARGET_ENV = env('TARGET_ENV', 'DEV')
+
+# Get aws EC2 instance metadata for logging, if available
+if bool_env(env('DJANGO_GET_EC2_METADATA', True)):
```
Dave, just for my clarity - you are not using the ec2 metadata yet, correct?

dtisza1 replied:
> Dave, just for my clarity - you are not logging the ec2 metadata yet, correct?

@oragame TY for reminding me to check. This is working. It defaults to True, but is set to False in docker-compose/web-env-vars.env when running in the local dev setup.

When testing via canary in the TEST env, this is what the firehose event JSON looks like:

{
  "time_of_event": "2021-09-14T14:13:02",
  "instance_id": "i-09ab7640b70a4c288",
  "image_id": "ami-019ff2daa7e8baac7",
  "component": "bb2.web",
  "vpc": "test",
  "log_name": "audit.global_state_metrics",
  "type": "global_state_metrics",
  "group_timestamp": "2021-09-14T14:13:01",
  "real_bene_cnt": 0,
  "synth_bene_cnt": 50,
  "global_apps_active_cnt": 17,
  "global_apps_inactive_cnt": 2,
  "global_apps_require_demographic_scopes_cnt": 17
}

oragame replied:
ok, 😎

oragame reviewed (approved):
Really nice work and methodology on getting this set up! I have successfully tested locally, and the Jenkins output looks just as I would expect.

JFU-GIT reviewed (approved):
Good work, approved

dtisza1 commented:
NOTE:  I am waiting until completing work for BB2-807/792, before merging this in. This is in case there are any small changes needed for the insights logging format.

dtisza1 commented:
I am closing out this PR.  

Logging will be funneled over to BFD-Insights using a log group subscription filter instead.  See BB2-918 for more details.

The docker-compose related refactoring work is split out in to the following PR:  https://github.com/CMSgov/bluebutton-web-server/pull/987
